### DISCUSSION TITLE
How can I do some additional cleanup when user uninstalls my MSIX app?

### DISCUSSION THREAD

**Discussion by JasonWei512** [microsoft/WindowsAppSDK#3061] [Q&A]
For example, my MSIX WinUI 3 app added some schedule tasks to Windows Task Scheduler. I want to delete these schedule tasks when user uninstalls my app. Is there any way to achieve this?

Relative discussion: https://techcommunity.microsoft.com/t5/msix/msix-custom-uninstall-processing/m-p/916781

**Comment by DrusTheAxe** [comment]
This is a current limitation. MSIX currently doesn't have a declarative option to define Windows Task Scheduler jobs (there's no `<Extension Category="windows.taskSchedule"...>`) so you've effectively scribbled outside of areas MSIX knows how to automagically cleanup on uninstall, and there's no open-ended hook for you to provide your own supplemental logic. Issue [Uninstall/Reset/Repair Tasks #2779](https://github.com/microsoft/WindowsAppSDK/issues/2779) is tracking a proposed solution.

MSIX supports [Background Tasks](https://learn.microsoft.com/en-us/windows/uwp/launch-resume/create-and-register-a-background-task) which are equivalent(ish) to Windows Task Scheduler jobs. Does that meet your needs?

**Comment by JasonWei512** [comment]
I want to run my app as admin at startup (silently, no UAC prompt). MSIX's Startup Task doesn't provide this feature.

The snipping tool [Snipaste](https://www.microsoft.com/store/productId/9P1WXPKB68KX) does this by creating a schedule task. However since MSIX doesn't allow developers to hook uninstalling process, after uninstalling Snipaste, the schedule task is still in Task Scheduler.

**Comment by DrusTheAxe** [comment]
>I want to run my app as admin at startup (silently, no UAC prompt). MSIX's Startup Task doesn't provide this feature.

If you define a windows.startupTask and the target app/exe has a Fusion (SxS) manifest declaring it runs elevated that's insufficient as you'd get a UAC prompt?

Likewise if you define `<Extension Category="windows.dataShortcut"...` pointing to `MyApp.lnk` which has the Run as Administrator checkbox checked and you include `MyApp.lnk` in your package that's also insufficient as you'd get a UAC prompt?

You're using a Task Scheduler job as a "StartupTask but launch as admin"?

If the windows.startupTask extension had an option `<RunAsAdministrator>true</>` to silently launch the app without the UAC prompt it would be equivalent to what you're doing via Task Scheduler?

Is "run as admin" a package-constant for you? If your appxmanifest.xml declared a windows.startupTask as `<RunAsAdministrator>true</>` and it was true at install time and couldn't be changed that fits your need? You're not looking to _optionally_ run at login?

If MSIX supported `<Extension Category="windows.taskSchedule" TrustLevel="elevated">...` where you could define your scheduled job via `appxmanifest.xml` (instead of code) would that meet your need? Do you need to be able to programmatically toggle this on and off post-installtime e.g. if you could call `TaskScheduler.Find("Snipaste").Enabled=false` to disable and `...=true` to enable would that cover it?

Trying to feel out the shape of your ask.

**Comment by DominicMaas** [comment]
Following this as I'm also interested in doing something like this (I have a Windows Service which I would like to cleanup), I took a brief look at `desktop6:UninstallAction`, but I see there has been a doc update explicitly mentioning that the app must be a "game" and packaged in a MSIXVC container, so making the assumption that this won't work (distributed outside MS Store)?

**Comment by DrusTheAxe** [comment]
>I have a Windows Service which I would like to cleanup

Is that a Packaged Service? i.e. [installed via MSIX](https://learn.microsoft.com/uwp/schemas/appxpackage/uapmanifestschema/element-desktop6-service). Those are automagically cleaned up when the package is removed just like other MSIX installed artifacts.

>I took a brief look at desktop6:UninstallAction, but...

Yes, that extension's restricted to packages in MSIXVC format. That's typically used by AAA games and poses several caveats and complexities making it ill suited to other ends.

PackagedServices are the recommended answer to provide a service via MSIX with all the usual benefits (including integrated automagic cleanup on uninstall).

**Reply by DominicMaas** [reply]
Sorry I should have been more specific, already using a packaged service installed via MSIX (which does get cleaned up), but the service will create a few folders on the system for state management, which ideally, I would clean-up on uninstall (but not upgrades).

**Comment by DrusTheAxe** [comment]
>the service will create a few folders on the system for state management

What kinds of folders where?

ApplicationData's created for your package family. Is there a problem using that?

**Reply by DominicMaas** [reply]
`C:\ProgramData\AppName`. Service runs as admin and supports multiple users on the system (talking to clients via pipes). Stores things like settings, logs, state, shared files for the program (which other programs will access).

**Reply by DrusTheAxe** [reply]
Sorry, I'm not aware of any storage automagically provided today to that end.

We've discussed it a few times -- the ability to register packages as LocalSystem and make ApplicationData available to them as a vFuture thing (not currently supported) e.g. stored under `C:\ProgramData\Packages\...pkgfamilyname...` as the moral equivalent to ApplicationData storage under `%USERPROFILE%\AppData\Local\Packages\...pkgfamilyname...` for packages registered to a user. And just liked packaged apps registered to a user, when the package is deregistered for the user the associated ApplicationData resources are deleted. Here'd it'd be when the Packaged Service is uninstalled which is when the package is removed for the last user, but same equivalent(ish) behavior.

I can think of a few ways how WinAppSDK could make that work (w/o changing Windows), so not just available on Windows vLatest but a downlevel-friendly solution. **If that interests you please create an Issue for consideration.**

I can think of a way you can accomplish your scenario for yourself, though it's a bit hacky. When you install your service and create `C:\ProgramData\AppName` also copy into it `DominicUninstallCleanup.exe`, create a scheduled job to run that every month/week/day(?) and make that exe do the cleanup `main() { RemoveDirectoryTree("C:\ProgramData\AppName") }`. Implement that in C/C++ and link with [/SWAPRUN](https://learn.microsoft.com/cpp/build/reference/swaprun-load-linker-output-to-swap-file?view=msvc-170) so you can delete that path including the exe itself (even if running).

Alternatively, you could create a long-running process whose purpose is to subscribe to PackageCatalog events and when your package is removed that long running process does the delete and quits (the long running exe is linked with /SWAPRUN so it can delete itself).

The polling approach is the distasteful part. It's effectively a form of garbage collection, but do it too often and you unnecessarily burn cycles+battery, do it too rarely and your 'garbage' wastes disk space that much longer. WinAppSDK can extend Windows in ways this can't to accomplish this more efficiently.

The long-running process is even less desirable as it's a permanent memory tax, not just an occasional run/check/nothing to do/quit hit.

**Reply by JosHuybrighs** [reply]
@DrusTheAxe 
These are interesting suggestions for doing some post-uninstall cleanup but require that the cleanup executable is copied to C:\ProgramData.
The problem with this I think, is that this requires elevation and that setting will probably not be allowed when I am distributing an App through the Microsoft Store and are not a Microsoft partner.

The best approach still would be a solution for [Uninstall/Reset/Repair Tasks #2779](https://github.com/microsoft/WindowsAppSDK/issues/2779). Will this come anytime soon?

I will need this since I also want to make use of the Windows Task Scheduler because of the current minimum of 15 minutes with a Timer-triggered Background task.

**Reply by DrusTheAxe** [reply]
> require that the cleanup executable is copied to C:\ProgramData.

Why?

If you're worried about "how does the exe delete itself or the dir it's in" that's what `/SWAPRUN` is for -- the exe is copied into the swapfile and executed from there (OK, more like the exe is paged into memory and run backed by the swapfile, not the original location). Thus c:\foo\bar.exe is not in use so it (and its containing directory) can be deleted without any pesky file-in-use errors. `/SWAPRUN` was invented (many moons ago) for uninstallers which is exactly your case.

You need to make sure your exe only depends on OS files (no imports to any other of your DLLs, doesn't need your other files, etc), or at least only depends on files you're not trying to delete.

**Reply by JosHuybrighs** [reply]
@DrusTheAxe 

The aproach I wanted to take for using the Windows Task Scheduler in my app (a folder sync/copy app, in a MSIX package) was to create entries in the scheduler for each job that needs scheduling. The 'action' part in each entry points to a small 'exec-cleanup' agent exe (written in C++ with /SWAPRUN) that either passes the execution request to my app or, in case the package in the mean time has been uninstalled, removes all scheduler entries.
If the exe isn't copied beforehand to a location outside of the app's installation folder, how can it then run after the app is uninstalled?
Maybe I don't completely understand how /SWAPRUN works but I assume that when the exe is executed Windows first copies the exe from its physical location on disk to a swapfile and then executes it from there.

**Reply by untoldbyte** [reply]
one thing to note


an msix packaged cant do a simple task like

create a read protected folder from end user to hide sensitive assets like a mov file/fonts. If user uninstalls the same app cannot be installed again.

thats a critical flaw.

either force delete automagically or allow custom uninstall hooks.


this limitation should also be called on the top of the docs

**Reply by DrusTheAxe** [reply]
>create a read protected folder from end user to hide sensitive assets like a mov file/fonts

Where are you creating this folder?

>either force delete automagically

If you create it under `ApplicationData` are you granting LocalSystem write access? If so then the deployment engine should be able to delete the 'protected folder' if/when the user uninstalls the package e.g.

```
sddl = "...(A;OICIIO;GA;;;SY)..."
sd = ConvertStringSecurityDescriptiorToSecurityDescriptor(sddl)
path = applicationdata.LocalFolder.Path + "\MyProtectedFolder"
CreateDirectory(path, sd)
```

(that's pseudo-code for the logic. Error handling, syntax for the APIs or alternative library equivalents, etc are left as an exercise for the reader)

The only reason there'd be a problem is if a file is in use and locked preventing deletion. In which case Windows does a best-effort deleting  what it can and try to cleanup the leftovers later, including after a reboot when files actively loaded and locked usually aren't.

Does this meet your need?
